gossipd/gossip_generation.c

```diff
@@ -307,48 +307,136 @@ void maybe_send_own_node_announce(struct daemon *daemon, bool startup)
+static u8 *channel_flags_access(const u8 *channel_update)
+{
+	/* BOLT #7:
+	 * 1. type: 258 (`channel_update`)
+	 * 2. data:
+	 *     * [`signature`:`signature`]
+	 *     * [`chain_hash`:`chain_hash`]
+	 *     * [`short_channel_id`:`short_channel_id`]
+	 *     * [`u32`:`timestamp`]
+	 *     * [`byte`:`message_flags`]
+	 *     * [`byte`:`channel_flags`]
+	 */
+	/* Note: 2 bytes for `type` field */
+	return cast_const(u8 *, &channel_update[2 + 64 + 32 + 8 + 4 + 1]);
+}
 
-	bool disable;
-	bool even_if_identical;
-	bool even_if_too_soon;
+static u8 *timestamp_access(const u8 *channel_update)
+{
+	/* BOLT #7:
+	 * 1. type: 258 (`channel_update`)
+	 * 2. data:
+	 *     * [`signature`:`signature`]
+	 *     * [`chain_hash`:`chain_hash`]
+	 *     * [`short_channel_id`:`short_channel_id`]
+	 *     * [`u32`:`timestamp`]
+	 *     * [`byte`:`message_flags`]
+	 *     * [`byte`:`channel_flags`]
+	 */
+	/* Note: 2 bytes for `type` field */
+	return cast_const(u8 *, &channel_update[2 + 64 + 32 + 8]);
+}
 
-	u16 cltv_expiry_delta;
-	struct amount_msat htlc_minimum, htlc_maximum;
-	u32 fee_base_msat, fee_proportional_millionths;
-};
+static bool is_disabled(const u8 *channel_update)
+{
+	return *channel_flags_access(channel_update) & ROUTING_FLAGS_DISABLED;
+}
 
-/* This generates a `channel_update` message for one of our channels.  We do
- * this here, rather than in `channeld` because we (may) need to do it
- * ourselves anyway if channeld dies, or when we refresh it once a week,
- * and so we can avoid creating redundant ones. */
-static void update_local_channel(struct local_cupdate *lc /* frees! */)
+static bool is_enabled(const u8 *channel_update)
 {
-	struct daemon *daemon = lc->daemon;
-	secp256k1_ecdsa_signature dummy_sig;
-	u8 *update, *msg;
-	u32 timestamp = gossip_time_now(daemon->rstate).ts.tv_sec, next;
-	u8 message_flags, channel_flags;
-	struct chan *chan = lc->local_chan->chan;
-	struct half_chan *hc;
-	const int direction = lc->local_chan->direction;
+	return !is_disabled(channel_update);
+}
 
-	/* Discard existing timer. */
-	lc->local_chan->channel_update_timer
-		= tal_free(lc->local_chan->channel_update_timer);
 
-	/* So valgrind doesn't complain */
-	memset(&dummy_sig, 0, sizeof(dummy_sig));
+static u32 timestamp_for_update(struct daemon *daemon,
+				const u32 *prev_timestamp,
+				bool disable)
+{
+	u32 timestamp = gossip_time_now(daemon->rstate).ts.tv_sec;
 
 	/* Create an unsigned channel_update: we backdate enables, so
 	 * we can always send a disable in an emergency. */
-	if (!lc->disable)
+	if (!disable)
 		timestamp -= GOSSIP_MIN_INTERVAL(daemon->rstate->dev_fast_gossip);
 
+	if (prev_timestamp) {
+		/* Timestamps can't go backwards! */
+		if (timestamp < *prev_timestamp)
+			timestamp = *prev_timestamp + 1;
+
+		/* If we ever use set-based propagation, ensuring the toggle
+		 * the lower bit in consecutive timestamps makes it more
+		 * robust. */
+		if ((timestamp & 1) == (*prev_timestamp & 1))
+			timestamp++;
```

Comment:
Seems like a premature optimisation (with no downside in deferring until it's needed), and may result in change in order if `prev_timestamp + 1 = timestamp`. This can potentially result in two elements flipping position in their order.

Reply:
This was actually code motion of existing code: agreed it was premature!  (It would be nice to have, since we propagate others' gossip, but we're a long way off).

It can't move backwards, since it's always set as minimum prev_timestamp + 1.